### PR TITLE
Planning reports

### LINKED ISSUE #8284: planner needs way to report events that it's waiting for
Copying from https://github.com/oxidecomputer/omicron/pull/8261#discussion_r2127369068:

> I am imagining that the we'll want to have an accumulator for "things that we're waiting for right now". Any time we bail out of a planning path (like we do here) because we're waiting on some condition, we'll want to append to that. Then I think we want to be able to report that with the upgrade status.

### LINKED ISSUE #8548: planner: adding new zones needs to know whether all lower-level updates are complete
This is closely related to #8284 and #8285, but those are phrased more in terms of "update"; this issue calls out that we have a similar problem when adding a new zone.

When the planner adds a new zone, it currently calls `image_source_for_new_zone()` (to determine the image source), which internally calls `is_zone_ready_for_update()` to decide whether to use the new or old TUF repo. `is_zone_ready_for_update()` has this TODO: https://github.com/oxidecomputer/omicron/blob/17e4c0e5b0b697c63bd05d267c0fdaca2b2e79d7/nexus/reconfigurator/planning/src/planner.rs#L1523-L1525

There are a number of cases where we might need to add a new zone in the middle of an upgrade, including at least:

* a new sled has been added (it will need NTP and crucible zones)
* a sled or disk was expunged, and it had services on it that now need to be restarted elsewhere to restore redundancy

In these cases, the planner needs to consider whether all RoT / SP / OS updates are complete to decide whether to start the new zone from the latest TUF repo or the previous one. I think this might be as "simple" as teaching `is_zone_ready_for_update()` how to check the results of `do_plan_mgs_updates()` which (after #8284) should be able to report whether all such updates are complete.

### PR DESCRIPTION
**Pull request by plotnick**
Fixes #8284, fixes #8548.

Replaces many (but not yet all) planner log messages with entries in a structured planning report. Those reports are comprised of "step reports", each of which corresponds to a subroutine of `do_plan`. The complete report may be formatted (in English) via a `Display` implementation. If planning from the `blueprint_planner` background task is successful, a report is attached to the `BlueprintPlannerStatus::Targeted` variant so that it may be displayed via `omdb`.

TODO, to be handled as follow-ups:
- [ ] Update for the rest of #8596, in particular `NoopConvertInfo`
- [ ] Update for #8456
- [x] Update for #8616
- [ ] Update for #8421

### CODE REVIEW

File: nexus/types/src/deployment/planning_report.rs [hunk 1]
```diff
@@ -0,0 +1,679 @@
...
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct PlanningNoopImageSourceStepReport {
+    pub no_target_release: bool,
+    pub skipped_sleds:
+        BTreeMap<SledUuid, PlanningNoopImageSourceSkipSledReason>,
+    pub skipped_zones:
+        BTreeMap<OmicronZoneUuid, PlanningNoopImageSourceSkipZoneReason>,
+    pub converted_zones: BTreeMap<SledUuid, (usize, usize)>,
```
Comment (jgallagher) [inline nexus/types/src/deployment/planning_report.rs]: Nit - can we turn `(usize, usize)` into a struct with field names? Just looking here, I don't know what these values are.

(Same for other `(usize, usize)` tuples below)

File: dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout [hunk 2]
```diff
@@ -37,20 +37,7 @@ generated inventory collection eb0796d5-ab8a-4f7b-a884-b4aeacb8ab51 from configu
...
-INFO sufficient ClickhouseServer zones exist in plan, desired_count: 0, current_count: 0
-INFO sufficient CockroachDb zones exist in plan, desired_count: 0, current_count: 0
-INFO sufficient CruciblePantry zones exist in plan, desired_count: 3, current_count: 3
-INFO sufficient InternalDns zones exist in plan, desired_count: 3, current_count: 3
-INFO sufficient ExternalDns zones exist in plan, desired_count: 3, current_count: 3
-INFO sufficient Nexus zones exist in plan, desired_count: 3, current_count: 3
-INFO sufficient Oximeter zones exist in plan, desired_count: 0, current_count: 0
 WARN cannot issue more SP updates (no current artifacts)
-INFO all zones up-to-date
-INFO will ensure cockroachdb setting, setting: cluster.preserve_downgrade_option, value: DoNotModify
```
Comment (jgallagher) [inline dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout:53]: Since we're losing all these logs, can `reconfigurator-cli blueprint-plan` emit the planning report? I think being able to see it in the CI diffs would also help a lot with reviewing the `Display` impls (and might give us ideas for thinks we should add).

File: nexus/reconfigurator/planning/src/planner.rs [hunk 3]
```diff
@@ -143,38 +147,57 @@ impl<'a> Planner<'a> {
...
     pub fn plan(mut self) -> Result<Blueprint, Error> {
-        self.check_input_validity()?;
-        self.do_plan()?;
+        let checked = self.check_input_validity()?;
+        self.do_plan(checked)?;
         Ok(self.blueprint.build())
     }
 
-    fn check_input_validity(&self) -> Result<(), Error> {
+    pub fn plan_and_report(
```
Comment (jgallagher) [inline nexus/reconfigurator/planning/src/planner.rs]: Do we need a separate method for this (as opposed to changing `plan()` to return the blueprint and report)? `do_plan` constructs the report anyway, so I'm not sure it's buying us much?

File: nexus/reconfigurator/planning/src/planner.rs [hunk 4]
```diff
@@ -513,11 +531,17 @@ impl<'a> Planner<'a> {
...
+        let mut report = PlanningNoopImageSourceStepReport::new();
+
+        let noop_info =
+            NoopConvertInfo::new(self.input, self.inventory, &self.blueprint)?;
+        noop_info.log_to(&self.log);
+
         let sleds = match noop_info {
             NoopConvertInfo::GlobalEligible { sleds } => sleds,
-            NoopConvertInfo::GlobalIneligible { .. } => return Ok(()),
+            NoopConvertInfo::GlobalIneligible { .. } => return Ok(report),
```
Comment (jgallagher) [inline nexus/reconfigurator/planning/src/planner.rs:594]: Can the report include some indication that no-op conversion is globally ineligible?

File: nexus/reconfigurator/planning/src/planner.rs [hunk 5]
```diff
@@ -959,30 +976,31 @@ impl<'a> Planner<'a> {
...
 
-    // Attempts to place `num_zones_to_add` new zones of `kind`.
-    //
-    // It is not an error if there are too few eligible sleds to start a
-    // sufficient number of zones; instead, we'll log a warning and start as
-    // many as we can (up to `num_zones_to_add`).
+    /// Attempts to place `num_zones_to_add` new zones of `kind`.
+    ///
+    /// It is not an error if there are too few eligible sleds to start a
+    /// sufficient number of zones; instead, we'll log a warning and start as
```
Comment (jgallagher) [inline nexus/reconfigurator/planning/src/planner.rs]:
```suggestion
    /// sufficient number of zones; instead, we'll report it and start as
```

File: nexus/reconfigurator/planning/src/planner.rs [hunk 6]
```diff
@@ -1095,24 +1109,45 @@ impl<'a> Planner<'a> {
...
+            report.waiting_on(ZoneUpdatesWaitingOn::DiscretionaryZones);
+            return Ok(report);
+        }
+
+        // ... or if there are still pending updates for the RoT / SP /
+        // Host OS / etc.
+        if mgs_updates.any_updates_pending() {
+            report.waiting_on(ZoneUpdatesWaitingOn::PendingMgsUpdates);
+            return Ok(report);
+        }
```
Comment (jgallagher) [inline nexus/reconfigurator/planning/src/planner.rs]: I think I'd move these checks back into `do_plan`? Definitely a judgment call on clarity, but I'm still swayed by Dave's arguments. I think it'd be nice if "skip this entire step" was obviously visible from `do_plan`. Even better if that means we can drop passing (some) earlier steps' reports into later steps' implementations (I know we'll still need some of these, but maybe not as many?).

File: nexus/reconfigurator/planning/src/planner.rs [hunk 7]
```diff
@@ -1095,24 +1109,45 @@ impl<'a> Planner<'a> {
...
+
+        // We are only interested in non-decommissioned sleds with
+        // running NTP zones (TODO: check time sync).
```
Comment (jgallagher) [inline nexus/reconfigurator/planning/src/planner.rs]: Is this new behavior, or is it moving something around?

File: dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout (hunk 2 above)
Comment (davepacheco) [reply dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout:53]: I like that idea.  It would also help in this PR to verify that the new reports include the same info that we had in the logs before.

File: dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout [hunk 8]
```diff
@@ -37,20 +37,7 @@ generated inventory collection eb0796d5-ab8a-4f7b-a884-b4aeacb8ab51 from configu
...
-INFO sufficient Clickhouse zones exist in plan, desired_count: 1, current_count: 1
-INFO sufficient ClickhouseKeeper zones exist in plan, desired_count: 0, current_count: 0
-INFO sufficient ClickhouseServer zones exist in plan, desired_count: 0, current_count: 0
-INFO sufficient CockroachDb zones exist in plan, desired_count: 0, current_count: 0
-INFO sufficient CruciblePantry zones exist in plan, desired_count: 3, current_count: 3
-INFO sufficient InternalDns zones exist in plan, desired_count: 3, current_count: 3
-INFO sufficient ExternalDns zones exist in plan, desired_count: 3, current_count: 3
-INFO sufficient Nexus zones exist in plan, desired_count: 3, current_count: 3
-INFO sufficient Oximeter zones exist in plan, desired_count: 0, current_count: 0
 WARN cannot issue more SP updates (no current artifacts)
```
Comment (davepacheco) [inline dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout]: I see that elsewhere some WARNs got removed, presumably in favor of a note in the report.  Should this one too?  (Fine if that's a follow-on PR.)

File: nexus/types/src/deployment/planning_report.rs [hunk 9]
```diff
@@ -0,0 +1,679 @@
...
+        }
+
+        for (sled_id, reason) in skipped_sleds.iter() {
+            writeln!(
+                f,
+                "* Skipping noop image source check on sled {sled_id}: {reason}"
+            )?;
+        }
+
+        // Very noisy in tests.
```
Comment (davepacheco) [inline nexus/types/src/deployment/planning_report.rs]: Hmm.  Is this useful or not?  If not, we should just remove it.

File: nexus/types/src/deployment/planning_report.rs [hunk 10]
```diff
@@ -0,0 +1,679 @@
...
+            Self::ArtifactNotInRepo { .. } => {
+                write!(f, "install dataset artifact hash not found in TUF repo")
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct PlanningMgsUpdatesStepReport {
+    pub pending_mgs_updates: PendingMgsUpdates,
```
Comment (davepacheco) [inline nexus/types/src/deployment/planning_report.rs:412]: I understand that this might be for a follow-up PR, I think I'd expect this to include a list of devices that remain to be updated but cannot be updated for some unexpected reason (missing image, missing inventory state for them, etc. -- these are the cases where `nexus_reconfigurator_planning::mgs_updates::try_make_update()` returns `None` after emitting a warning).  That in turn will let us really fix #8285.

We might also want this to include:
- a list of updates that were determined to be completed
- a list of updates that were determined to be impossible and cleared
- a list of updates that we configured

I say "list" -- currently it can only be one, but the code doesn't assume that because it's a policy choice we've made for safety, not a limitation of the implementation.

This sounds like exactly the stuff @jgallagher suggests could go in a follow on PR (which makes sense to me).

File: nexus/reconfigurator/planning/src/planner.rs [hunk 11]
```diff
@@ -209,15 +232,10 @@ impl<'a> Planner<'a> {
...
+                // up when we ask for commissioned sleds? Report, but don't try to
                 // decommission it again.
                 (SledPolicy::Expunged, SledState::Decommissioned) => {
-                    error!(
-                        self.log,
-                        "decommissioned sled returned by \
-                         SledFilter::Commissioned";
-                        "sled_id" => %sled_id,
-                    );
+                    report.zombie_sleds.push(sled_id);
```
Comment (davepacheco) [inline nexus/reconfigurator/planning/src/planner.rs:291]: This is a nice example of the improvement this API brings!

File: nexus/reconfigurator/planning/src/planner.rs [hunk 12]
```diff
@@ -651,14 +669,13 @@ impl<'a> Planner<'a> {
...
+                )?,
             )? == Ensure::Added
             {
-                info!(
-                    &self.log,
-                    "found sled missing NTP zone (will add one)";
-                    "sled_id" => %sled_id
-                );
+                report.sleds_missing_ntp_zone.insert(sled_id);
                 self.blueprint.record_operation(Operation::AddZone {
```
Comment (davepacheco) [inline nexus/reconfigurator/planning/src/planner.rs:730]: Do we think this new mechanism would eventually replace `record_operation()`?  (I didn't previously know about that.)

File: nexus/reconfigurator/planning/src/reports.rs [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    PlanningReport, PlanningZoneUpdatesStepReport,
+};
+use omicron_uuid_kinds::BlueprintUuid;
+
+/// A blueprint planning report minus the blueprint ID that the
+/// report is for. Returned by [`crate::planner::Planner::do_plan`]
+/// when all planning steps are complete, but before the blueprint
+/// has been built (and so we don't yet know its ID).
+#[derive(Debug)]
+pub(crate) struct InterimPlanningReport {
```
Comment (davepacheco) [inline nexus/reconfigurator/planning/src/reports.rs]: It looks like the blueprint_id is assigned when we call `BlueprintBuilder::build()`.  That can only be called once.  I wonder if we could remove this whole type (and this file) by generating the blueprint id when we call `BlueprintBuilder::new_based_on()` instead and making it available to the caller via a `new_blueprint_id()` method.  Not a big deal either way.

**Review by davepacheco — COMMENTED** [review]
I like this!  I agree with @jgallagher that it makes sense to land this sooner rather than later and iterate in follow-on PRs.  I don't have any blockers here.  (I also didn't review the mechanics in planner.rs and planning_report.rs with a fine-tooth comb.)

File: nexus/types/src/deployment/planning_report.rs (hunk 1 above)
Comment (plotnick) [reply nexus/types/src/deployment/planning_report.rs]: It seems that `JsonSchema` doesn't really like tuples either, so they were all removed in 3fa1a21e5.

File: dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout (hunk 2 above)
Comment (plotnick) [reply dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout:53]: Thanks, f0aabe55b adds the report to the end of blueprint display, and also to the message returned by `reconfigurator-cli::cmd_blueprint_plan`.

File: nexus/reconfigurator/planning/src/planner.rs (hunk 3 above)
Comment (plotnick) [reply nexus/reconfigurator/planning/src/planner.rs]: f0aabe55b moves the report into the blueprint itself, so the need for this method goes away.

File: nexus/reconfigurator/planning/src/planner.rs (hunk 4 above)
Comment (plotnick) [reply nexus/reconfigurator/planning/src/planner.rs:594]: It definitely can and should, but I haven't converted the `image_source` module yet (that's the second TODO in the description above), and so I'd rather hold off on that and do it all at once as a follow-up.

File: nexus/reconfigurator/planning/src/planner.rs (hunk 5 above)
Comment (plotnick) [reply nexus/reconfigurator/planning/src/planner.rs]: Thanks, fixed in 6d5b56487.

File: nexus/reconfigurator/planning/src/planner.rs (hunk 6 above)
Comment (plotnick) [reply nexus/reconfigurator/planning/src/planner.rs]: Sure, done in 6d5b56487.

File: nexus/reconfigurator/planning/src/planner.rs (hunk 7 above)
Comment (plotnick) [reply nexus/reconfigurator/planning/src/planner.rs]: I was confused about #8353; reverted in 6d5b56487. I also had some other bugs in my translation of how various NTP edge-cases were handled, which were fixed in bd0b6d56b.

File: dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout (hunk 8 above)
Comment (plotnick) [reply dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout]: I haven't updated the `mgs_updates` module for planning reports yet. I think I could take this with the follow-up for #8285.

File: nexus/types/src/deployment/planning_report.rs (hunk 9 above)
Comment (plotnick) [reply nexus/types/src/deployment/planning_report.rs]: It doesn't seem so to me, but I lack context on its importance. Should we remove the commented-out printing only, or the field too?

File: nexus/types/src/deployment/planning_report.rs (hunk 10 above)
Comment (plotnick) [reply nexus/types/src/deployment/planning_report.rs:412]: That all sounds right to me, and I agree it should be a follow-up.

File: nexus/reconfigurator/planning/src/planner.rs (hunk 12 above)
Comment (plotnick) [reply nexus/reconfigurator/planning/src/planner.rs:730]: Good question; I don't know. But I noticed the similarity, too.

File: nexus/reconfigurator/planning/src/reports.rs (hunk 13 above)
Comment (plotnick) [reply nexus/reconfigurator/planning/src/reports.rs]: We can indeed! Done in 48dce2a62. It had a larger surface area than I would have liked because we now need to pass the `PlannerRng` in to the builder, and so also to `Planner::new_based_on`. This touched a lot of tests, but didn't really change anything.

File: dev-tools/reconfigurator-cli/tests/output/cmds-noop-image-source-stdout [hunk 14]
```diff
@@ -165,23 +165,20 @@ INFO install dataset artifact hash not found in TUF repo, ignoring for noop chec
...
-INFO will ensure cockroachdb setting, setting: cluster.preserve_downgrade_option, value: DoNotModify
 generated blueprint 58d5e830-0884-47d8-a7cd-b2b3751adeb4 based on parent blueprint 8da82a8e-bf97-4fbd-8ddd-9f6462732cf1
+Planning report for blueprint 58d5e830-0884-47d8-a7cd-b2b3751adeb4:
+Chicken switches:
+    add zones with mupdate override:   false
+
+* Noop converting 6/6 install-dataset zones to artifact store on sled 98e6b7c2-2efa-41ca-b20a-0a4d61102fe6
+* Noop converting 5/6 install-dataset zones to artifact store on sled aff6c093-197d-42c5-ad80-9f10ba051a34
+* 1 pending MGS update:
+  * model0:serial0: Sp { expected_active_version: ArtifactVersion("0.0.1"), expected_inactive_version: NoValidVersion }
```
Comment (davepacheco) [inline dev-tools/reconfigurator-cli/tests/output/cmds-noop-image-source-stdout]: At first I thought: "it's probably worth adding the new version here", then I thought "well, this isn't supposed to be a summary of everything done -- that's the blueprint itself.  Maybe it should have even less?"  I don't have a strong feeling about this.

File: dev-tools/reconfigurator-cli/tests/output/cmds-set-mgs-updates-stdout [hunk 15]
```diff
@@ -422,6 +425,9 @@ parent:    ad97e762-7bf1-45a6-a98f-60afb7e491c0
     sled      2      model2        serial2         e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855   1.1.0              Sp { expected_active_version: ArtifactVersion("1.0.0"), expected_inactive_version: Version(ArtifactVersion("1.0.1")) }
 
 
+Nothing to report on planning for blueprint cca24b71-09b5-4042-9185-b33e9f2ebba0.
```
Comment (davepacheco) [inline dev-tools/reconfigurator-cli/tests/output/cmds-set-mgs-updates-stdout]: What does this message mean?  Does it mean the blueprint is the same as the parent?

edit: I think it might also mean "we loaded this from the database and so have no report for it".  Maybe this should say "Empty planning report ..."?  That might avoid confusing someone who sees this and thinks there have been no changes, when it might just be that we loaded this from the database.

File: nexus/db-queries/src/db/datastore/deployment.rs [hunk 16]
```diff
@@ -1683,6 +1684,9 @@ impl DataStore {
             )?;
         }
 
+        // FIXME: Once reports are stored in the database, read them out here.
```
Comment (davepacheco) [inline nexus/db-queries/src/db/datastore/deployment.rs:1314]: Can you file an issue for this?

File: nexus/types/src/deployment/planning_report.rs [hunk 17]
```diff
@@ -601,26 +835,37 @@ impl ZoneUpdatesWaitingOn {
...
 
-#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+#[derive(
+    Clone, Debug, Deserialize, Serialize, PartialEq, Eq, Diffable, JsonSchema,
+)]
+#[serde(rename_all = "snake_case", tag = "type")]
 pub enum ZoneUnsafeToShutdown {
-    Cockroachdb(CockroachdbUnsafeToShutdown),
+    Cockroachdb { reason: CockroachdbUnsafeToShutdown },
+    BoundaryNtp { total_boundary_ntp_zones: usize, synchronized_count: usize },
```
Comment (davepacheco) [inline nexus/types/src/deployment/planning_report.rs:917]: I think InternalDNS might belong here now.

**Review by davepacheco — APPROVED** [review]
Looking good!  Thanks, and sorry for my slowness here.

File: dev-tools/reconfigurator-cli/tests/output/cmds-set-mgs-updates-stdout (hunk 15 above)
Comment (plotnick) [reply dev-tools/reconfigurator-cli/tests/output/cmds-set-mgs-updates-stdout]: The intent here is to represent a report that is "empty" in the sense of "devoid of interesting content", and in particular that won't display anything. Previously these would display as just the header:
```
Planning report for blueprint {blueprint_id}:
```
I wanted to indicate the empty case explicitly, rather than just not printing anything below the header.

Successful planning runs with nontrivial diffs can result in these if nothing unusual or unexpected occurred. But see above re: tension; it's not clear not me whether the report should be "standalone", in which case we probably don't really want empty reports.

Regardless, "Empty planning report" is indeed more precise, wording updated in dcac3dd69.

File: dev-tools/reconfigurator-cli/tests/output/cmds-noop-image-source-stdout (hunk 14 above)
Comment (plotnick) [reply dev-tools/reconfigurator-cli/tests/output/cmds-noop-image-source-stdout]: I think what to put in the report vs just in the blueprint (or its diff) is going to be a continuing tension. Since a report is (currently) tied to a particular planning run, they could be difficult to interpret without their blueprint (diff); on the other hand, we might want them to be more "standalone" so they can be interpreted by a developer or operator without requiring (possibly expunged) blueprints alongside them. We could decide as a matter of policy which way we'd prefer to go here, or just experiment and see what's useful and what's not.

File: nexus/db-queries/src/db/datastore/deployment.rs (hunk 16 above)
Comment (plotnick) [reply nexus/db-queries/src/db/datastore/deployment.rs:1314]: Sure: #8788.

File: nexus/types/src/deployment/planning_report.rs (hunk 17 above)
Comment (plotnick) [reply nexus/types/src/deployment/planning_report.rs:917]: Right, as of #8683. Updated in 370c959d7.

**Comment by plotnick** [comment]
@jgallagher: You were right (as usual): leaving out a planning step report for `do_plan_mupdate_override` and just doing what it and `do_plan_noop_image_source` currently want with the `NoopConvertInfo` made the merge straightforward. The cleanup effort was, as predicted, tricky, and I will have to come back to it next week.

File: dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout [hunk 18]
```diff
@@ -37,21 +37,15 @@ generated inventory collection eb0796d5-ab8a-4f7b-a884-b4aeacb8ab51 from configu
...
 WARN cannot issue more MGS-driven updates (no current artifacts)
-INFO all zones up-to-date
-INFO will ensure cockroachdb setting, setting: cluster.preserve_downgrade_option, value: DoNotModify
 generated blueprint 8da82a8e-bf97-4fbd-8ddd-9f6462732cf1 based on parent blueprint dbcbd3d6-41ff-48ae-ac0b-1becc9b2fd21
+Planning report for blueprint 8da82a8e-bf97-4fbd-8ddd-9f6462732cf1:
+Chicken switches:
+    add zones with mupdate override:   false
+
+* No zpools in service for NTP zones on sleds: 00320471-945d-413c-85e7-03e091a70b3c
+* Discretionary zone placement waiting for NTP zones on sleds: 00320471-945d-413c-85e7-03e091a70b3c
```
Comment (sunshowers) [inline dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout]: nit: could you make these start with lower case to match the other outputs we produce?

File: nexus/reconfigurator/planning/src/planner.rs [hunk 19]
```diff
@@ -189,55 +182,71 @@ impl<'a> Planner<'a> {
...
-            {
-                self.do_plan_zone_updates()?;
-            }
-        }
+            self.do_plan_add(&mgs_updates)?
+        } else {
+            PlanningAddStepReport::waiting_on(
+                ZoneAddWaitingOn::MupdateOverrides,
+            )
+        };
```
Comment (sunshowers) [inline nexus/reconfigurator/planning/src/planner.rs]: can we record the results of both `plan_mupdate_override_res.is_empty()` and the `add_zones_with_mupdate_override` chicken switch here? one of the things I tried to do was to add logging for when we would have ordinarily have not added zones, but the chicken switch made us do it anyway -- it can be confusing to figure out why a decision was made.

File: nexus/reconfigurator/planning/src/planner.rs (hunk 12 above)
Comment (sunshowers) [reply nexus/reconfigurator/planning/src/planner.rs:730]: I think this makes sense yeah -- record_operation just records when an operation was affirmatively performed, which seems like a subset of what the report does.

**Review by sunshowers — APPROVED** [review]
A few comments but nothing blocking landing this -- can be addressed in followups if you think it makes sense.

File: dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout (hunk 18 above)
Comment (plotnick) [reply dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout]: sure, done in 7822a53c2.

File: nexus/reconfigurator/planning/src/planner.rs (hunk 19 above)
Comment (plotnick) [reply nexus/reconfigurator/planning/src/planner.rs]: no problem, done in d02feafef.
